### PR TITLE
feat/file-api

### PR DESCRIPTION
**Pull request by Kerwin-Jhong**
Fixes #

### CODE REVIEW

File: lambda-assets/files/create-category/app.ts [hunk 1]
```diff
@@ -8,15 +11,38 @@ export async function handler(event: { [key: string]: any }) {
     const validated = request.validate(joi => {
       return {
         name: joi.string().required(),
+        parentId: joi.string().allow(null),
+        description: joi.string(),
```
Comment (MinCheTsai) [inline lambda-assets/files/create-category/app.ts:15]: allow null?

File: .projenrc.js [hunk 2]
```diff
@@ -23,10 +23,14 @@ const project = new AwsCdkConstructLibrary({
   bundledDeps: [
     'joi',
     'semver',
+    'md5',
```
Comment (MinCheTsai) [inline .projenrc.js:26]: ?

File: .projenrc.js [hunk 3]
```diff
@@ -23,10 +23,14 @@ const project = new AwsCdkConstructLibrary({
...
+    'lodash',
     'uuid',
     '@types/node@15.12.2',
     '@types/uuid',
     '@types/semver',
+    '@types/md5',
```
Comment (MinCheTsai) [inline .projenrc.js:32]: ?

File: lambda-assets/files/create-category/app.ts [hunk 4]
```diff
@@ -8,15 +12,42 @@ export async function handler(event: { [key: string]: any }) {
...
+          TableName: `${CATEGORY_TABLE_NAME}`,
+          Item: {
+            categoryId: md5.update(name).digest('hex'),
+            name,
+            description: request.input('description'),
+            createdAt: timestamp,
+            updatedAt: timestamp,
+          },
+          ...parameters
+      })
```
Comment (MinCheTsai) [inline lambda-assets/files/create-category/app.ts:45]:
```
let itemParameters: { [key: string]: any } = {};
if (request.has('parentId')) {
  itemParameters.categoryId = md5.update(`${parentId}-${name}`).digest('hex');
  itemParameters.parentId = parentId;
} else {
  itemParameters.categoryId = md5.update(name).digest('hex')
}
await ddbDocClient.send(
  new PutCommand({
    TableName: `${CATEGORY_TABLE_NAME}`,
    Item: {
      ...itemParameters,
      name,
      description: request.input('description'),
      createdAt: timestamp,
      updatedAt: timestamp,
    },
});
```

File: lambda-assets/files/create-file/app.ts [hunk 5]
```diff
@@ -1,26 +1,46 @@
...
+    await ddbDocClient.send(
+      new PutCommand({
+        TableName: `${FILE_TABLE_NAME}`,
+        Item: {
+          fileId: request.input('checksum'),
+          version: request.input('version'),
+          categoryId: request.input('categoryId'),
+          location: request.input('location'),
+          description: request.input('description'),
+          createdAt: Date.now(),
```
Comment (MinCheTsai) [inline lambda-assets/files/create-file/app.ts:35]: currentTime

File: lambda-assets/files/delete-category/app.ts [hunk 6]
```diff
@@ -1,15 +1,27 @@
...
+const { CATEGORY_TABLE_NAME } = process.env;
+
 export async function handler(event: { [key: string]: any }) {
   const request = new Request(event);
   const response = new Response();
   try {
+    const ddbDocClient = DynamoDBDocumentClient.from(
+      new DynamoDBClient({})
+    );
+    await ddbDocClient.send(new DeleteCommand({
```
Comment (MinCheTsai) [inline lambda-assets/files/delete-category/app.ts:14]: new line

File: lambda-assets/files/list-categories/app.ts [hunk 7]
```diff
@@ -1,14 +1,60 @@
...
+      return response.json({
+        data: categories,
+        nextToken: LastEvaluatedKey,
+      });
+    } else {
+      parameters.ExpressionAttributeNames = {
+        '#parentId': 'parentId',
+      };
+      parameters.FilterExpression = 'attribute_not_exists(#parentId)';
+      const { Items: categories, LastEvaluatedKey } = await ddbDocClient.send(
```
Comment (MinCheTsai) [inline lambda-assets/files/list-categories/app.ts:46]: LastEvaluatedKey: nextToken

File: lambda-assets/files/list-files-by-category/app.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+const { FILE_TABLE_NAME } = process.env;
+
+export async function handler(event: { [key: string]: any }) {
+  const request = new Request(event);
+  const response = new Response();
+  try {
+    const ddbDocClient = DynamoDBDocumentClient.from(
+      new DynamoDBClient({})
+    );
+    const { Items: files, LastEvaluatedKey } = await ddbDocClient.send(
```
Comment (MinCheTsai) [inline lambda-assets/files/list-files-by-category/app.ts:14]: LastEvaluatedKey: nextToken

File: lambda-assets/files/list-files-by-category/app.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        TableName: `${FILE_TABLE_NAME}`,
+        IndexName: 'query-by-category-id',
+        KeyConditionExpression: '#categoryId = :categoryId',
+        ExpressionAttributeNames: {
+          '#categoryId': 'categoryId',
+        },
+        ExpressionAttributeValues: {
+          ':categoryId': request.parameter('categoryId'),
+        },
+        ExclusiveStartKey: request.get('nextToken', undefined),
```
Comment (MinCheTsai) [inline lambda-assets/files/list-files-by-category/app.ts:25]: undefined will error

File: lambda-assets/files/list-files/app.ts [hunk 10]
```diff
@@ -1,16 +1,27 @@
...
+const { FILE_TABLE_NAME } = process.env;
+
 export async function handler(event: { [key: string]: any }) {
   const request = new Request(event);
   const response = new Response();
   try {
+    const ddbDocClient = DynamoDBDocumentClient.from(
+      new DynamoDBClient({})
+    );
+    const { Items: files, LastEvaluatedKey } = await ddbDocClient.send(
```
Comment (MinCheTsai) [inline lambda-assets/files/list-files/app.ts:14]: LastEvaluatedKey: nextToken

File: lambda-assets/files/list-files/app.ts [hunk 11]
```diff
@@ -1,16 +1,27 @@
...
+        TableName: `${FILE_TABLE_NAME}`,
+        ExclusiveStartKey: request.get('nextToken', undefined),
+      })
+    );
     return response.json({
-      categoryId: request.parameter('categoryId'),
-      files: [],
-      nextToken: request.get('nextToken', undefined),
+      files,
+      nextToken: LastEvaluatedKey,
```
Comment (MinCheTsai) [inline lambda-assets/files/list-files/app.ts:22]: undefined will error

File: lambda-assets/files/update-file/app.ts [hunk 12]
```diff
@@ -1,15 +1,36 @@
...
+import { DynamoDBDocumentClient, UpdateCommand } from '@aws-sdk/lib-dynamodb';
 import { Request, Response } from '../../utils';
 
+const { FILE_TABLE_NAME } = process.env;
+
 export async function handler(event: { [key: string]: any }) {
   const request = new Request(event);
   const response = new Response();
   try {
+    const ddbDocClient = DynamoDBDocumentClient.from(
```
Comment (MinCheTsai) [inline lambda-assets/files/update-file/app.ts:11]: validate?

Comment (MinCheTsai) [reply lambda-assets/files/update-file/app.ts:11]: if 404?

File: test/lambda/file.test.ts [hunk 13]
```diff
@@ -1,55 +1,396 @@
...
+      expectedCategory,
+    ],
+    nextToken: {
+      categoryId: '48cc48e0d55bfd83114031498f21d640',
+    },
+  },
+};
+
+const expectedFile = {
+  DBData: {
```
Comment (MinCheTsai) [inline test/lambda/file.test.ts:57]: DBData?

File: test/lambda/file.test.ts [hunk 14]
```diff
@@ -1,55 +1,396 @@
...
+    updatedAt: 'Test',
+  },
+  //: expectedInvalidJob,
+  invalidDBError: <AwsError>{
+    Code: 'ResourceNotFoundException',
+  },
+};
+
+test('Create category success', async () => {
+  const ddbMock = mockClient(DynamoDBDocumentClient);
```
Comment (MinCheTsai) [inline test/lambda/file.test.ts:75]: ddbMockClient

File: test/lambda/file.test.ts [hunk 15]
```diff
@@ -1,55 +1,396 @@
...
+  ddbMock.on(PutCommand).resolves;
```
Comment (MinCheTsai) [inline test/lambda/file.test.ts:76]: ddbMock.on(PutCommand, {??}).resolves;

Comment (MinCheTsai) [reply test/lambda/file.test.ts:76]: process.env.TABLE_NAME = 'xxx'

File: lambda-assets/files/create-file/app.ts [hunk 16]
```diff
@@ -1,26 +1,47 @@
...
   const response = new Response();
   try {
     const validated = request.validate(joi => {
       return {
-        categoryId: joi.string().required(),
         location: joi.string().uri().required(),
-        version: joi.string().required(), // validate semanticVersion
-        checksumType: joi.string().allow(['md5', 'crc32', 'sha1']),
         checksum: joi.string().required(),
+        checksumType: joi.string().allow('md5', 'crc32', 'sha1'),
```
[the remark this answers is not in the document]
Comment (liaojason2softchef) [reply lambda-assets/files/create-file/app.ts]: 註解解釋嗎？

File: lambda-assets/files/list-categories/app.ts [hunk 17]
```diff
@@ -1,14 +1,59 @@
...
-      nextToken: request.get('nextToken', undefined),
-    });
+    const ddbDocClient = DynamoDBDocumentClient.from(
+      new DynamoDBClient({})
+    );
+    const parentId = request.get('parentId', undefined);
+    let parameters: { [key: string]: any } = {};
+    if (request.has('nextToken')) {
+        parameters.ExclusiveStartKey = {
+        Key: request.get('nextToken'),
```
Comment (MinCheTsai) [inline lambda-assets/files/list-categories/app.ts:18]: indent?

Comment (MinCheTsai) [reply lambda-assets/files/list-categories/app.ts:18]: nextToken需要轉換

File: lambda-assets/files/create-file/app.ts [hunk 18]
```diff
@@ -1,26 +1,48 @@
...
-        version: joi.string().required(), // validate semanticVersion
-        checksumType: joi.string().allow(['md5', 'crc32', 'sha1']),
-        checksum: joi.string().required(),
+        checksumType: joi.string().allow('md5', 'crc32', 'sha1'), 
+        checksum: joi.string().when('checksumType', { is: 'md5', then: joi.string().length(32).required()})
+        .concat(joi.string().when('checksumType', { is: 'crc32', then: joi.string().length(8).required()}))
+        .concat(joi.string().when('checksumType', { is: 'sha1', then: joi.string().length(40).required()})),
+        version: joi.string().required(),
+        categoryId: joi.string().required(),
+        //checkMD5Length: joi.string().when('checksumType', { is: 'md5', then: joi.string().length(32).required})
```
Comment (MinCheTsai) [inline lambda-assets/files/create-file/app.ts:18]: clean code

File: lambda-assets/files/list-categories/app.ts [hunk 19]
```diff
@@ -1,14 +1,72 @@
...
+        ).toString('base64')
+      }
+      return response.json({
+        categories,
+        nextToken,
+      });
+    };
   } catch (error) {
-    return response.json(error);
+    console.log(error)
```
Comment (MinCheTsai) [inline lambda-assets/files/list-categories/app.ts:69]: clean code

File: lambda-assets/files/update-category/app.ts [hunk 20]
```diff
@@ -1,15 +1,41 @@
...
+        },
+        UpdateExpression: "set #description = :description",
+        ExpressionAttributeNames: {
+          '#description': 'description'
+        },
+        ExpressionAttributeValues: {
+          ":description": request.input('description', ''),
+        },
+      })
+    );
```
Comment (MinCheTsai) [inline lambda-assets/files/update-category/app.ts:34]: 404?

File: src/file-api.ts [hunk 21]
```diff
@@ -83,6 +83,11 @@ export class FileApi extends cdk.Construct {
           httpMethod: HttpMethod.GET,
           lambdaFunction: this.createListFilesFunction(),
         },
+        {
+          path: '/files/{categoryId}',
```
Comment (MinCheTsai) [inline src/file-api.ts:87]: /categories/{categoryId}/files

File: lambda-assets/files/create-file/app.ts [hunk 22]
```diff
@@ -1,26 +1,47 @@
...
+      new DynamoDBClient({})
+    );
+    const currentTime = Date.now();
+    await ddbDocClient.send(
+      new PutCommand({
+        TableName: process.env.FILE_TABLE_NAME,
+        Item: {
+          fileId: request.input('checksum'),
+          version: request.input('version'),
+          categoryId: request.input('categoryId'),
```
Comment (MinCheTsai) [inline lambda-assets/files/create-file/app.ts:33]: if category id not exists?

File: lambda-assets/files/create-file/app.ts [hunk 23]
```diff
@@ -1,26 +1,59 @@
...
+    const { Item: category } = await ddbDocClient.send(
+      new GetCommand({
+        TableName: process.env.CATEGORY_TABLE_NAME,
+        Key: {
+          categoryId,
+        },
+      })
+    );
+    if (!category) {
+      return response.error('Not found.', 404);
```
Comment (MinCheTsai) [inline lambda-assets/files/create-file/app.ts:36]: 422

Comment (MinCheTsai) [reply lambda-assets/files/create-file/app.ts:36]: 建議比照輸入錯誤辦理

**Review by MinCheTsai — APPROVED** [review]

**Pull request merged**